src/graphics/pyro/pyro.h

```diff
@@ -65,67 +63,47 @@ class CPyro
     friend class CPyroManager;
 
     //! Creates pyrotechnic effect
-    bool        Create(PyroType type, CObject* obj, float force);
+    bool        Create();
     //! Destroys the object
     void        DeleteObject();
 
+    virtual void AfterCreate();
+    virtual void UpdateEffect();
+    virtual void AfterEnd();
+
 public:
-    CPyro(); // should only be called by CPyroManager
-    ~CPyro();
+    CPyro(CObject *obj);
+    virtual ~CPyro();
 
     //! Indicates whether the pyrotechnic effect is complete
-    Error       IsEnded();
+    virtual Error IsEnded();
 
     //! Indicates that the object binds to the effect no longer exists, without deleting it
     void        CutObjectLink(CObject* obj);
 
     //! Management of an event
-    bool        EventProcess(const Event& event);
+    virtual bool EventProcess(const Event& event);
 
 protected:
-    //! Displays the error or eventual information
-    //! Information can be linked to the destruction of an insect, a vehicle or building
-    void        DisplayError(PyroType type, CObject* obj);
-
     //! Creates light to accompany a pyrotechnic effect
     void        CreateLight(Math::Vector pos, float height);
     //! Removes the binding to a pyrotechnic effect
     void        DeleteObject(bool primary, bool secondary);
 
     //! Creates an explosion with triangular form of particles
-    void        CreateTriangle(CObject* obj, ObjectType oType, int part);
-
-    //! Starts the explosion of a vehicle
-    void        ExploStart();
-    //! Ends the explosion of a vehicle
-    void        ExploTerminate();
-
-    //! Starts a vehicle fire
-    void        BurnStart();
-    //! Adds a part move
-    void        BurnAddPart(int part, Math::Vector pos, Math::Vector angle);
-    //! Advances of a vehicle fire
-    void        BurnProgress();
-    //! Indicates whether a part should be retained
-    bool        BurnIsKeepPart(int part);
-    //! Ends the fire of an insect or a vehicle
-    void        BurnTerminate();
-
-    //! Start of an object freight falling
-    void        FallStart();
-    //! Seeks an object to explode by the falling ball of bees
-    CObject*    FallSearchBeeExplo();
-    //! Fall of an object's freight
-    void        FallProgress(float rTime);
-    //! Indicates whether the fall is over
-    Error       FallIsEnded();
+    void        CreateTriangle(CObject* obj, ObjectType oType, int part,
```

Comment:
It is a bad idea to create a function with loads of parameters especially default parameters. It's very difficult to see which arguments at the call site correspond to which parameters.
Alternatives:
1. Make `maxHParticleSpeed`, `maxVParticleSpeed`, `minParticleMass`, `maxParticleMass` and `overridePercent` instance fields of `CPyro` and set them in the constructors of the descendants
2. Similar to 2 except make them virtual functions
3. [Use struct wrappers or proxy objects](https://softwareengineering.stackexchange.com/a/329733/386495)